corehq/apps/userreports/sql.py

```diff
@@ -30,12 +31,18 @@ def save(self, doc):
             table = self.get_table()
             for indicator_row in indicator_rows:
                 with self.engine.begin() as connection:
-                    # delete all existing rows for this doc to ensure we aren't left with stale data
-                    delete = table.delete(table.c.doc_id == doc['_id'])
-                    connection.execute(delete)
-                    all_values = {i.column.id: i.value for i in indicator_row}
-                    insert = table.insert().values(**all_values)
-                    connection.execute(insert)
+                    try:
+                        # delete all existing rows for this doc to ensure we aren't left with stale data
+                        delete = table.delete(table.c.doc_id == doc['_id'])
+                        connection.execute(delete)
+                        all_values = {i.column.id: i.value for i in indicator_row}
+                        insert = table.insert().values(**all_values)
+                        connection.execute(insert)
+                    except IntegrityError:
+                        # Someone beat us to it. Concurrent inserts can happen
+                        # when a doc is processed by the celery rebuild task
+                        # at the same time as the pillow.
+                        pass
```

Comment:
This may not perform very well because of the many transactions and connections involved. I assume it's the insert that would raise `IntegrityError` (you could narrow the `try` block to only include that statement). Would it make sense to put a where condition on the insert, so it does nothing but still succeeds when someone beat us to it? If you did that it seems like you could do all of these inserts with a single transaction/connection.